src/storage/mod.rs

```diff
@@ -0,0 +1,98 @@
+use thiserror::Error;
+
+mod nats_kv;
+mod state;
+
+use semver::Version;
+use async_trait::async_trait;
+
+use state::{
+    LatticeState,
+    LatticeStateBuilder,
+    LatticeId,
+};
+
+/////////////
+// Engines //
+/////////////
+
+pub type EngineName = String;
+
+/// An engine for performing some task that is pluggable
+pub trait Engine {
+    /// Get the name of the lattice storage
+    fn name() -> EngineName;
+
+    /// Get the version of the lattice storage
+    fn version() -> Version;
+}
+
+/////////////////////
+// Storage Engines //
+/////////////////////
+
+/// Errors that are related to storage engines
+#[derive(Debug, Error)]
+pub enum StorageEngineError {
+    /// When the underlying engine throws an error
+    #[error("Storage engine encountered an error: {0}")]
+    Engine(String),
+
+    #[error("Storage engine [{0}] not connected")]
+    NotConnected(EngineName),
+
+    /// When an operation is not implemented
+    #[error("Operation not implemented: {0}")]
+    NotImplemented(String),
+
+    /// An error when the cause is unknown. This happens very rarely and should be considered fatal
+    #[error("Unknown error has occured")]
+    Unknown,
+}
+
+/// Storage engine
+#[async_trait]
+pub trait StorageEngine: Engine {
+    /// Connect the storage engine
+    async fn connect(&mut self) -> Result<(), StorageEngineError>;
+
+    /// Initialize the storage engine (if necessary)
+    async fn init(&mut self) -> Result<(), StorageEngineError>;
+}
+
+/////////////
+// Storage //
+/////////////
+
+/// Errors that are related to storage
+#[derive(Debug, Error)]
+pub enum StorageError {
+    /// Errors that arise at the storage layer
+    #[error("Storage error: {0}")]
+    StorageError(String),
+
+    /// Errors that arise at the network layer
+    #[error("Network error: {0}")]
+    NetworkError(String),
+
+    /// An error when the cause is unknown. This happens very rarely and should be considered fatal
+    #[error("Unknown error has occured: {0}")]
+    Unknown(String),
+}
+
+/// A trait that indicates the ability of a struct to store state
+/// Objects that implement Stores can save some given State, identifiable by StateId, and updatable with the given StateBuilder
+#[async_trait]
+pub trait Stores<State, StateId, StateBuilder> {
+    // Get a particular piece of state
+    async fn get(id: StateId) -> Result<State, StorageError>;
+
+    // Store a piece of state
+    async fn store(state: State) -> Result<State, StorageError>;
+
+    // Update an existing piece of state
+    async fn update(id: StateId, update: StateBuilder) -> Result<(), StorageError>;
+}
```

Comment:
Missing delete?

Reply:
Yup! it'll be in there soon, don't worry, just wanted to get something up review ASAP

Reply:
[added!](https://github.com/wasmCloud/wadm/pull/62/commits/617e1d78d1849fc8ea129048b2fac6ac086cf38c)